### PR TITLE
add internal problem type and internal user type

### PR DESCRIPTION
**Pull request by iv404**
only internal user and admin can see the internal problem.

### CODE REVIEW

File: trunk/src/main/java/cn/edu/uestc/acmicpc/service/impl/ProblemServiceImpl.java [hunk 1]
```diff
@@ -40,6 +41,16 @@ public ProblemServiceImpl(ProblemDao problemDao) {
         problemCondition.getCondition());
   }
 
+  @SuppressWarnings("unchecked")
+  @Override
+  public List<Integer> getAllVisibleNormalProblemIds() throws AppException {
```
Comment (lyhypacm) [inline trunk/src/main/java/cn/edu/uestc/acmicpc/service/impl/ProblemServiceImpl.java]: I prefer merge all usages into a method, i.e. set visible and type as parameters.

File: trunk/src/main/java/cn/edu/uestc/acmicpc/db/condition/impl/StatusCondition.java [hunk 2]
```diff
@@ -103,6 +103,9 @@ public StatusCondition() {
   @Exp(mapField = "problemByProblemId.isVisible", type = ConditionType.EQUALS)
   public Boolean isVisible;
```
Comment (lyhypacm) [inline trunk/src/main/java/cn/edu/uestc/acmicpc/db/condition/impl/StatusCondition.java]: isProblemVisible

File: trunk/src/main/java/cn/edu/uestc/acmicpc/db/condition/impl/StatusCondition.java [hunk 3]
```diff
@@ -103,6 +103,9 @@ public StatusCondition() {
...
 
+  @Exp(mapField = "problemByProblemId.type", type = ConditionType.EQUALS)
+  public Integer type;
```
Comment (lyhypacm) [inline trunk/src/main/java/cn/edu/uestc/acmicpc/db/condition/impl/StatusCondition.java]: problemType

Comment (iv404) [inline trunk/src/main/java/cn/edu/uestc/acmicpc/db/condition/impl/StatusCondition.java]: code updated

**Comment by iv404** [comment]
code updated.

File: trunk/src/main/java/cn/edu/uestc/acmicpc/db/condition/impl/StatusCondition.java [hunk 4]
```diff
@@ -100,12 +102,35 @@ public StatusCondition() {
...
+   *
+   * @see ProblemType
+   */
+  @Exp(mapField = "problemByProblemId.type", type = Condition.ConditionType.EQUALS)
+  public ProblemType problemType;
 
   @Override
   public Condition getCondition() throws AppException {
-    Condition condition = super.getCondition();
+    Condition preCondition = super.getCondition();
```
Comment (mzry1992) [inline trunk/src/main/java/cn/edu/uestc/acmicpc/db/condition/impl/StatusCondition.java]: Can you expose a common util to share this part with ProblemCondition?

File: trunk/src/test/sql/uestcojtest.sql [hunk 5]
```diff
@@ -87,6 +87,7 @@ CREATE  TABLE IF NOT EXISTS `uestcojtest`.`problem` (
   `javaMemoryLimit` INT NOT NULL DEFAULT 65535 ,
   `dataCount` INT NOT NULL DEFAULT 0 ,
   `difficulty` INT NOT NULL DEFAULT 1 ,
+  `type` INT NOT NULL DEFAULT 0 ,
```
Comment (mzry1992) [inline trunk/src/test/sql/uestcojtest.sql]: Also add this line in  `trunk/src/main/sql/uestcoj.sql`

And create a patch sql in `trunk/src/main/sql/patches/` for developer

**Comment by iv404** [comment]
sql file updated.button to edit type of problem added.

**Pull request merged**